DAOS-10037 cart: Improve the abort code

liw opened the pull request:
There are a few abort problems in two future use cases.

  - In a daos_engine, the SWIM xstream aborts all in-flight RPCs sent by
    all xstreams to a specific rank. This will be implemented by https://github.com/daos-stack/daos/pull/8426.
    When called by the SWIM xstream, crt_req_abort may
      - introduce data races on crt_rpc_priv fields with the xstream
	owning the RPC;
      - race with crt_req_send_internal, crt_hg_req_send_cb, etc., and
	is incomplete in the handling of certain RPC states;
      - invoke the RPC completion callback on the SWIM xstream instead
	of the xstream owning the RPC.

  - In dfuse, multiple libdaos application threads send, progress, and
    abort RPCs in the same context. We might introduce abort calls to
    implement dfuse operation-level timeouts or faster shutdowns. The
    first two problems under the daos_engine use case also apply here.

This patch addresses those problems in the following manner.

  - Add crt_rpc_priv.crp_mutex to prevent data races on crt_rpc_priv
    fields.
      - Extend and fix the existing fine-to-coarse lock order to prevent
	deadlocks. See the comment on crt_rpc_priv.crp_mutex.
      - Remove the recursions in the
	crt_req_send_internal-crt_req_hg_add_lookup path.
      - Remove crp_on_wire, which is superseded by crp_mutex and
	crp_state.
      - Remove the unused RPC_STATE_REPLY_RECEVED.

  - Use crt_context.cc_bh_timeout to avoid calling RPC completion
    callbacks in undesirable threads. (We could also introduce a new
    queue tracking RPCs to be aborted.)
      - Set crt_rpc_priv.crp_timeout_ts to 0 in crt_req_abort so that
	the progress thread will complete the RPC. This will return
	-DER_TIMEDOUT instead of -DER_CANCELED for an aborted RPC. It is
	not an issue for daos since daos already treats -DER_CANCELED as
        a network error that shall be retried.

  - Print which RPCs can't be aborted in crt_ctx_epi_abort. This has
    been helpful during the development.

  - Remove the recursion in crt_context_req_untrack.

Signed-off-by: Li Wei <wei.g.li@intel.com>
Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/1/execution/node/1124/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented on src/cart/crt_internal_types.h:
```diff
@@ -215,6 +216,17 @@ struct crt_context {
 	char			 cc_self_uri[CRT_ADDR_STR_MAX_LEN];
 };
 
+#define CTX_ADDREF(ctx)										\
```
(style)  Single statement macros should not use a do {} while (0) loop

daosbuild1 commented on src/cart/crt_internal_types.h:
```diff
@@ -215,6 +216,17 @@ struct crt_context {
+	atomic_fetch_add(&ctx->cc_refcount, 1);							\
+} while (0)
+
+#define CTX_DECREF(ctx)										\
```
(style)  Single statement macros should not use a do {} while (0) loop

daosbuild1 reviewed (changes requested):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-11030/4/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/4/execution/node/145/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/5/execution/node/1131/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/8/execution/node/1132/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/9/execution/node/1099/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/10/execution/node/859/log

daosbuild1 commented:
Test stage Functional Hardware Medium UCX Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/10/execution/node/1135/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/10/execution/node/1119/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage NLT on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/11/execution/node/638/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/13/execution/node/940/log

daosbuild1 commented:
Test stage Functional Hardware Medium UCX Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/13/execution/node/956/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/13/execution/node/898/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Build on Leap 15.4 with Intel-C and TARGET_PREFIX completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/14/execution/node/365/log

daosbuild1 commented:
Test stage Build on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/14/execution/node/372/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/14/execution/node/342/log

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/14/execution/node/327/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.4 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/14/execution/node/330/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/15/execution/node/941/log

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/15/execution/node/957/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Large completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/16/execution/node/846/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

liw commented:
@frostedcmos, @liuxuezhao, could you let me know if the approach I use in this patch is acceptable or not, please? I'm not sure about the change on when to start the RPC timer.

daosbuild1 commented:
Test stage Functional Hardware Medium completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-11030/17/testReport/(root)/

liw commented:
The test failure also happens with other PRs and therefore is unrelated.

liuxuezhao commented on src/cart/crt_context.c:
```diff
@@ -1202,50 +1362,49 @@ crt_context_req_untrack(struct crt_rpc_priv *rpc_priv)
+		D_MUTEX_LOCK(&epi->epi_mutex);
+		if (tmp_rpc->crp_state == RPC_STATE_QUEUED && credits_available(epi) > 0) {
+			tmp_rpc->crp_state = RPC_STATE_INITED;
 
```
original calling of "crt_req_timeout_track" removed, not quite sure you mean it is unnecessary?

liw replied:
Yes, this patch changes crt_context_req_track to call crt_req_timeout_track for both waiting and inflight requests. Hence, the waiting request here should already be in the timeout binheap.

liuxuezhao replied:
ok, original code will not add to timeout binheap if it is queued, but seems not bad to track timeout for QUEUED case also.

liw replied:
It does have some downside: I guess we didn't track timeouts for waiting requests because we wanted to avoid timeouts due to local waits. I'm also a little concerned if we will see such timeouts as a result of this patch.

On the other hand, tracking timeouts for waiting requests makes it possible for users to limit the time spent on waiting for an RPC. For example, in the future, if we track the deadline of an incoming RPC in an engine, and the RPC handler needs to send an other RPC to a different engine, then we may want to get a -DER_TIMEDOUT for the outgoing RPC before the deadline of the incoming RPC.

Personally, I'm not sure which way is _definitely_ better, and would be happy to revert this change if you guys prefer the current behavior.

liw replied:
@frostedcmos also commented on this change. Since everyone, including myself, turns out to be a bit concerned, I'm working on restoring the previous behavior.

liw replied:
Restored to the previous behavior (i.e., not starting timer until inflight).

liw commented on src/cart/crt_context.c:
```diff
@@ -902,6 +1037,15 @@ crt_req_timeout_hdlr(struct crt_rpc_priv *rpc_priv)
+		RPC_ERROR(rpc_priv, "aborting waiting to group %s, rank %d, tgt_uri %s\n",
+			  grp_priv->gp_pub.cg_grpid, tgt_ep->ep_rank, rpc_priv->crp_tgt_uri);
+		D_MUTEX_LOCK(&rpc_priv->crp_epi->epi_mutex);
+		d_list_del_init(&rpc_priv->crp_epi_link);
```
I think you're right that we need to RPC_DECREF here. I've modified crt_context_req_untrack a little (otherwise it can't handle QUEUED request) and used it here.

liw reviewed (commented):
Thanks, Xuezhao.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Medium UCX Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/18/execution/node/1055/log

liuxuezhao reviewed (approved):
LGTM, there are a few test failures, not very sure if it is related although.

liw commented:
> LGTM, there are a few test failures, not very sure if it is related although.

Yeah, I've been investigating the rebuild_rc failure. It appears that the progress ULT on the system xstream of rank 1 was blocked by some Argobots lock or eventual, because that xstream had not been able to receive any RPCs since while it could still run other ULTs.

liw commented:
> > LGTM, there are a few test failures, not very sure if it is related although.
> 
> Yeah, I've been investigating the rebuild_rc failure. It appears that the progress ULT on the system xstream of rank 1 was blocked by some Argobots lock or eventual, because that xstream had not been able to receive any RPCs since, while it could still run other ULTs.

This turns out to be a deadlock regression. I've added a fix commit "Fix URI lookup deadlocks".

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/19/execution/node/1116/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

frostedcmos reviewed (commented):
looks good to me, just need to go more thoroughly again to understand all the changes. mostly minor things inline

daosbuild1 commented:
Test stage Functional Hardware Medium UCX Provider completed with status FAILURE.  https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-11030/20/display/redirect

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

liw commented:
> Test stage Functional Hardware Medium UCX Provider completed with status FAILURE. https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-11030/20/display/redirect

Looks like some `scp` error caused this failure...

frostedcmos commented:
"Start the RPC timer before crt_req_send returns, rather than
afterwards when the queued RPC moves to the inflight queue, so
that the an RPC is always in crt_context.cc_bh_timeout."

I am concerned about this particular change, as the timer now starts for all wait-queued rpcs as well. Let's assume scenario where server takes 6 seconds to process an RPC, timeout is set to 10, and inflight depth is 1. Client asynchronously submitting two rpcs to the same target will now result in a second rpc timing out, while previous behavior would allow both rpcs to complete.  At the same time client submitting 2 rpcs synchronously would still work.

liw commented on src/cart/crt_hg.c:
```diff
@@ -1318,7 +1314,7 @@ crt_hg_req_send(struct crt_rpc_priv *rpc_priv)
 		rpc_priv->crp_state = RPC_STATE_FWD_UNREACH;
 	} else {
-		rpc_priv->crp_on_wire = 1;
+		rpc_priv->crp_state = RPC_STATE_REQ_SENT;
```
Yes, this is the only place where `crp_state` becomes `REQ_SENT`. With the addition of `crp_mutex`, we no longer need to set `REQ_SENT` before calling `crt_hg_req_send`, and thus no longer need the `crp_on_wire` flag.

liw commented on src/cart/crt_context.c:
```diff
@@ -458,6 +460,64 @@ crt_rpc_complete(struct crt_rpc_priv *rpc_priv, int rc)
+	uint32_t	  p_len;
+};
+
+static int
```
Moved to gurt as d_vec_pointers.

liw commented on src/cart/crt_context.c:
```diff
@@ -458,6 +460,64 @@ crt_rpc_complete(struct crt_rpc_priv *rpc_priv, int rc)
+		uint32_t	  cap;
+
+		if (pointers->p_cap == 0)
+			cap = 8;
```
This is not a magic number; its meaning is clear. Moreover, the number is only referenced here once. I prefer not to introduce a macro scenarios like this.

liw commented on src/cart/crt_context.c:
```diff
@@ -458,6 +460,64 @@ crt_rpc_complete(struct crt_rpc_priv *rpc_priv, int rc)
+		if (pointers->p_cap == 0)
+			cap = 8;
+		else
+			cap = 2 * pointers->p_cap;
```
AFAIK doubling is the most common strategy.

liw commented on src/cart/crt_context.c:
```diff
@@ -710,36 +803,83 @@ int
+	 * orders, we hold at most one of them at a time.
+	 */
+
+	rc = pointers_init(&ctxs, 8 /* cap */);
```
The number depends on the type of the objects we store in this instance of the vector; I'm afraid using a common macro might discourage people from considering what a suitable number should be for each vector.

daosbuild1 commented on src/cart/crt_context.c:
```diff
@@ -1135,25 +1233,31 @@ crt_context_req_track(struct crt_rpc_priv *rpc_priv)
-	d_list_t		 submit_list;
-	struct crt_rpc_priv	*tmp_rpc;
-	int			 rc;
+	struct crt_ep_inflight	*epi = rpc_priv->crp_epi;;
```
(style)  Statements terminations use 1 semicolon

daosbuild1 replied:
```suggestion
	struct crt_ep_inflight	*epi = rpc_priv->crp_epi;
```

daosbuild1 commented on src/cart/crt_context.c:
```diff
@@ -1196,58 +1302,88 @@ crt_context_req_untrack(struct crt_rpc_priv *rpc_priv)
+			D_MUTEX_UNLOCK(&crt_ctx->cc_mutex);
+			if (rc != 0)
+				RPC_ERROR(tmp_rpc,
+					"crt_req_timeout_track failed, rc: %d.\n", rc);
```
```suggestion
					  "crt_req_timeout_track failed, rc: %d.\n", rc);
```

daosbuild1 reviewed (dismissed):
Style warning(s) for job https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-11030/22/
Please review https://wiki.hpdd.intel.com/display/DC/Coding+Rules

daosbuild1 commented:
Test stage checkpatch completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/22/execution/node/145/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/22/execution/node/342/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

liw commented:
- Restored the previous timeout semantics (i.e., RPC timers do not start until RPCs are in flight).
- Released crp_mutex before calling RPC completion callback, to avoid deadlock concerns. Renamed code that locks/unlocks crp_mutex to maintain readability: crt_rpc_lock, crt_rpc_unlock, crt_rpc_complete_and_unlock.
- Addressed Alex's comments.

daosbuild1 commented:
Test stage Functional Hardware Medium Verbs Provider completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/23/execution/node/1239/log

frostedcmos reviewed (changes requested):
Need a bit more time to review but in a meantime, change in context_destroy (noted inline) is needed.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

liw reviewed (commented):
Thanks, Alex.

frostedcmos commented on src/cart/crt_rpc.c:
```diff
@@ -1344,14 +1349,22 @@ crt_req_send_internal(struct crt_rpc_priv *rpc_priv)
 		}
 		break;
 	case RPC_STATE_ADDR_LOOKUP:
+send_immediately:
 		rc = crt_req_send_immediately(rpc_priv);
```
not sure i like the label in the middle of the case statement. can we just create the outside label and jump there instead?

frostedcmos reviewed (commented):
Looks good, 2 comments inline

daosbuild1 commented:
Test stage Functional Hardware Medium UCX Provider completed with status UNSTABLE.  https://build.hpdd.intel.com/job/daos-stack/job/daos//view/change-requests/job/PR-11030/25/testReport/(root)/

liw commented on src/cart/crt_context.c:
```diff
@@ -635,39 +700,9 @@ crt_context_destroy(crt_context_t crt_ctx, int force)
-	D_MUTEX_DESTROY(&ctx->cc_mutex);
-	D_DEBUG(DB_TRACE, "destroyed context (idx %d, force %d)\n", ctx->cc_idx, force);
-	D_FREE(ctx);
+	CTX_DECREF(ctx);
```
I don't have a good idea, only an average one: Instead of calling `CTX_DEFREF(ctx)`, we wait for the reference count of `ctx` to drop to 1, then call `crt_context_free`.

liw commented on src/cart/crt_context.c:
```diff
@@ -635,39 +700,9 @@ crt_context_destroy(crt_context_t crt_ctx, int force)
-
-	d_binheap_destroy_inplace(&ctx->cc_bh_timeout);
-
 	D_MUTEX_UNLOCK(&ctx->cc_mutex);
```
Noticed a bug in this patch: `crt_context_destroy` acquires `cc_mutex` and `epi_mutex` in the wrong order, and `crt_ctx_epi_abort` should not release and acquire `cc_mutex`.

liw replied:
Fixed.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/26/execution/node/311/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.4 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/26/execution/node/358/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/26/execution/node/280/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

daosbuild1 commented:
Test stage Build RPM on EL 8 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/27/execution/node/306/log

daosbuild1 commented:
Test stage Build RPM on Leap 15.4 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/27/execution/node/324/log

daosbuild1 commented:
Test stage Build DEB on Ubuntu 20.04 completed with status FAILURE.  https://build.hpdd.intel.com//job/daos-stack/job/daos/view/change-requests/job/PR-11030/27/execution/node/321/log

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.

frostedcmos commented on src/cart/crt_context.c:
```diff
@@ -616,45 +606,73 @@ crt_ctx_epi_abort(d_list_t *rlink, void *arg)
+out_epis:
+	for (i = 0; i < epis.p_len; i++) {
+		struct crt_ep_inflight *epi = epis.p_buf[i];
+
+		d_hash_rec_decref(&ctx->cc_epi_table, &epi->epi_link);
```
should this be wrapped in D_MUTEX_LOCK/UNLOCK on ctx->cc_mutex?

liw replied:
Indeed. Fixed.

liw commented on src/cart/crt_context.c:
```diff
@@ -762,19 +808,26 @@ crt_rank_abort(d_rank_t rank)
+	 * will not be destroyed. We don't block if the lock is busy, because a
+	 * crt_context_destroy call may take the lock for quite some time.
+	 */
+	rc = pthread_rwlock_trywrlock(&crt_context_destroy_lock);
```
Added D_RWLOCK_TRYWRLOCK.

daosbuild1 reviewed (commented):
LGTM.  No errors found by checkpatch.
